### PR TITLE
Space artillery

### PR DESCRIPTION
**Pull request by Qulibly**
## About the PR

It is meant to add Space Artillery, a weapon permitting security to much more effectively chase and take down pirates.
While also allowing true ship to ship combat.

**To Do list**
- [X] Create YAML for the entity to serve as foundation for the weapon and allow first testing
- [ ] Add ability to fire the weapon without need to hold it in hand
- [ ] Proper sprites and sounds
- [ ] QoL changes to make the use and firing of the weapon more fun
- [ ] Creation of dedicated shells (Ammunition like EMP shell, HE shell etc)
- [ ] Balance revision
- [ ] Codifying board, construction costs, research required and/or availability of it

### CODE REVIEW

**Comment by Qulibly** [comment]
I'm personally wondering how I should do the firing part. One idea is to try and use the raising of events to try and simulate the firing.

As currently. Space Artillery works as basically one very big [Admeme] Improvised Pneumatic Cannon.
And using code from that.

**Comment by Kesiath** [comment]
Make a new morgue type “loading bay” that one thing can be inserted into at a time. For sprites, I was thinking maybe more something along the lines of the particle accelerator, it’s multiple parts. You’d need the breach (the reused morgue), and a barrel. If using air to fire the rounds instead of gunpowder, the breach should also need an air intake pipe and a pressure cylinder.

**Comment by Qulibly** [comment]
The Space Artillery is meant to work as simple entity. Without being modular.
The "Particle accelerator like" are meant to be the SGAs, Ship Grade Armament. Those will be modular.
I want Space Artillery to be something as simple to work as possible so that it could lay groundwork for more complex things.

As for operating.
It doesnt actually use any gas. It generally is meant to use internal structure and gunpowder to fire.
And making it have inventory to load will probably make it much more easy to use.
Even on the fly, like being pulled by someone.

**Comment by Qulibly** [comment]
> How about making weapons that automatically fire on "armed and possesive active ship weaponry"? Like emp field that disable for time shuttle if they get too close or real artillery, look on npc update they've got some pretty interesting changes.

When it comes to that I believe it would be extremely complex to make and could be rather difficult.
Not only in tracking aspect, but also choosing WHAT it should target. 
Additionally NPC code is rather difficult.
Moreover aiming said artillery in space won't be a very easy thing since shuttles will be moving alot AND also sideways movement relative to where you're aiming is transferring the velocity to the projectile itself, complicating the aiming even more.
So the automatic targeting weapon could have extreme inaccurancy even if made working.

And I personally feel like making it automatic target something would be out of the scope of this PR

**Comment by Peptide90** [comment]
Upstream are getting some forms of ship weapons currently but this PR seems pretty good. Is it still being worked on / any media for it?

**Comment by Qulibly** [comment]
> Upstream are getting some forms of ship weapons currently but this PR seems pretty good. Is it still being worked on / any media for it?

Yes still working on it.
The code currently is in really solid place and soon the core features should have foundation finished.

As for media. I could get screenshots in but they wouldn't show much and things can still change.
For videos I am sadly limited in how long video/enough quality I can record. I had some recordings earlier but then my recording software borked. Will need to figure out OBS and see if can get some using it.

And yes saw the upstream ship weapons. They only have the remote activated kind. Though for Frontier they would need to be completely rebalanced.
But the sprites are definetly amazing.

Speaking of which.
**At moment biggest issue is lack of SPRITES and SOUNDS**
Which while for now I'm focusing on code and mechanics. Eventually its going to become the main thing remaining to do.

**Comment by Qulibly** [comment]
I have also ran quite a few playtests with random folk. Doing battles, testing stuff etc.

Though I did had quite alot of delays due to being sick and exams.
But work should be getting more back on track.

**Comment by Peptide90** [comment]
> > Upstream are getting some forms of ship weapons currently but this PR seems pretty good. Is it still being worked on / any media for it?
> 
> Yes still working on it. The code currently is in really solid place and soon the core features should have foundation finished.
> 
> As for media. I could get screenshots in but they wouldn't show much and things can still change. For videos I am sadly limited in how long video/enough quality I can record. I had some recordings earlier but then my recording software borked. Will need to figure out OBS and see if can get some using it.
> 
> And yes saw the upstream ship weapons. They only have the remote activated kind. Though for Frontier they would need to be completely rebalanced. But the sprites are definetly amazing.
> 
> Speaking of which. **At moment biggest issue is lack of SPRITES and SOUNDS** Which while for now I'm focusing on code and mechanics. Eventually its going to become the main thing remaining to do.

I recommend shareX for recording short videos or gifs. Very lightweight and easy to use. Otherwise a lot of people use GeForce experience if you're on nvidia.

**Comment by VividPups** [comment]
> > > Upstream are getting some forms of ship weapons currently but this PR seems pretty good. Is it still being worked on / any media for it?
> > 
> > 
> > Yes still working on it. The code currently is in really solid place and soon the core features should have foundation finished.
> > As for media. I could get screenshots in but they wouldn't show much and things can still change. For videos I am sadly limited in how long video/enough quality I can record. I had some recordings earlier but then my recording software borked. Will need to figure out OBS and see if can get some using it.
> > And yes saw the upstream ship weapons. They only have the remote activated kind. Though for Frontier they would need to be completely rebalanced. But the sprites are definetly amazing.
> > Speaking of which. **At moment biggest issue is lack of SPRITES and SOUNDS** Which while for now I'm focusing on code and mechanics. Eventually its going to become the main thing remaining to do.
> 
> I recommend shareX for recording short videos or gifs. Very lightweight and easy to use. Otherwise a lot of people use GeForce experience if you're on nvidia. Just for an artillery PR I'd expect some videos of them firing and causing damage 👍

Medal is another one. its the one im currently doing.. which i have some grapits about but its pretty good

**Comment by Qulibly** [comment]
> @Qulibly this PR needs tended to soonish i think, it's been stale for a while and need to merge master into this at the very least.. what is your plan for this PR ?

Im still working on getting the core of all armament systems working

Its honestly rather big feature with quite some complexity
To add to fact it sisnt help that was sick alot and then had college delaying.


As for goals.
Right now im working on getting the armament activation system usable for player.

And for that I need to figure out how to construct UI with console.
Then there will be left is one more system to not make destructable armaments feel cheap.

Some sprites for it are being worked on but still ia gonna take a while.

In short im on roadblock.

Current woeking systems:
- armament firing
- remotely activated armaments
- mountable armaments
- space recoil (both for armaments and boosters)
- initial armament balance(will need to revision it once syatems are present)
- armament activation with kicking out of cloak and turning vessel color to red is already working (though lacking proper way to activate it ingame unless I link it to armament safety)
- Frontier Outpost safezone safety (armament shells cant accidentally damage Frontier itself)

**Comment by Qulibly** [comment]
@GreaseMonk Gotcha!
When it comes at moment.
I personally could time/meeting where coukd try to understand UI structure and how to code it so it works properly.
Cause at moment still don't have an idea how to make it.
Only got potential concept/goal.
And reverse engineering already made consoles aint the easiest way to learn UI.

**Comment by Vonsant** [comment]
@Qulibly need some help with this?

**Comment by Qulibly** [comment]
> @Qulibly need some help with this?

Could use some sure. For now it has been on giant hiatus due to college exams, but will have time to resume soon.
The code for armaments themselves and activation is done and recovered

currently I need to get the copilot console with UI, BUI and code for it designed, coded adn tested to work

**Comment by Vonsant** [comment]
@Qulibly what about sprites? What did you need?
